pkg/dhcp6client/client.go

```diff
@@ -0,0 +1,148 @@
+// Package dhcp6client implements a simple DHCPv6 client.
+//
+// It only supports soliciting and receiving the initial DHCP address with few
+// options.
+package dhcp6client
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/mdlayher/dhcp6"
+	"github.com/mdlayher/eui64"
+	"golang.org/x/net/ipv6"
+)
+
+const (
+	listenPort = 546
+	dstPort    = 547
+)
+
+var (
+	// All DHCP servers and relay agents on the local network segment (RFC 3315)
+	// IPv6 Multicast (RFC 2464)
+	// insert the low 32 Bits of the multicast IPv6 Address into the Ethernet Address (RFC 7042 2.3.1.)
+	multicastMAC = net.HardwareAddr([]byte{0x33, 0x33, 0x00, 0x01, 0x00, 0x02})
+)
+
+var (
+	errTimedOut = errors.New("timed out")
+)
+
+// Client is a simple DHCPv6 client.
+type Client struct {
+	// The HardwareAddr to send in the request.
+	srcMAC net.HardwareAddr
+
+	// Packet socket to send on.
+	conn *ipv6.PacketConn
+
+	// Max number of attempts to multicast DHCPv6 solicits.
+	// -1 means infinity.
+	retry int
+
+	// Timeout for each Solicit try.
+	timeout time.Duration
+}
+
+func New(haddr net.HardwareAddr, t time.Duration, r int) (*Client, error) {
+	ip, err := eui64.ParseMAC(net.ParseIP("::"), haddr)
+	if err != nil {
+		return nil, err
+	}
+
+	// If this doesn't work like this, we may have to SO_BINDTODEVICE to
+	// the interface. If that doesn't work... then we're back to raw
+	// sockets. Hope not.
+	c, err := net.ListenPacket("udp6", fmt.Sprintf("%s:%d", ip, listenPort))
+	if err != nil {
+		return nil, err
+	}
+	return &Client{
+		srcMAC:  haddr,
+		conn:    ipv6.NewPacketConn(c),
+		timeout: t,
+		retry:   r,
+	}, nil
+}
+
+func (c *Client) Solicit() (*dhcp6.Packet, error) {
+	solicitPacket, err := newSolicitPacket(c.srcMAC)
+	if err != nil {
+		return nil, fmt.Errorf("new solicit packet: %v", err)
+	}
+
+	var packet *dhcp6.Packet
+	// Each retry takes the amount of timeout at worst.
+	for i := 0; i < c.retry || c.retry < 0; i++ {
+		packet, err = c.MulticastPacket(solicitPacket)
+		switch err {
+		case errTimedOut:
```

Comment:
this is gonna be a bit hazardous with large numbers of clients as there is no randomization in the retry time. It maybe ok, but in the old days this could cause request storms as nodes doing multiple requests with the same timeout can self-synchrnonize.